internal/tuf/v01/root.go

```diff
@@ -723,6 +723,41 @@ func (p *PropagationDirective) GetDownstreamPath() string {
 	return p.DownstreamPath
 }
 
+// setter added to update exsisting directive
+func (p *PropagationDirective) SetUpstreamRepository(upstreamRepository string) {
+	p.UpstreamRepository = upstreamRepository
+}
+
+func (p *PropagationDirective) SetUpstreamReference(upstreamReference string) {
+	p.UpstreamReference = upstreamReference
+}
+
+func (p *PropagationDirective) SetDownstreamReference(downstreamReference string) {
+	p.DownstreamReference = downstreamReference
+}
+
+func (p *PropagationDirective) SetDownstreamPath(downstreamPath string) {
+	p.DownstreamPath = downstreamPath
+}
+
+// Updatepropagation checks if there is a duplicate or update and deal with that
+func (r *RootMetadata) UpdatePropagationDirective(directive tuf.PropagationDirective) error {
+	// find if there is a duplicate(exact same) or update(same name, different in other)
+	// in the slice in linear search, then update the found propagation
+	for i := range r.Propagations {
+		if r.Propagations[i].GetName() == directive.GetName() {
+			r.Propagations[i] = directive
+			return nil
+		}
+	}
+
+	//if the directive is new, call add propagation directive to deal with it
+	if err := r.AddPropagationDirective(directive); err != nil {
+		return err
+	}
+	return nil
```

Comment:
If the loop above completes without returning then we can just assume that the directive was not found and return an error as such.

```suggestion
return ErrPropagationDirectiveNotFound
```